### ISSUE TITLE
Message consumer should indicate a closed connection

### ISSUE THREAD

**Issue by jeremyherbert** [eclipse/paho.mqtt.cpp#131]
Hi, not sure if this is an issue with the C++ library or the C one (I suspect it is the C one). The following code reliably deadlocks within a few minutes of running in my Ubuntu 16.04 environment. I just run a mosquitto server on localhost and then have three threads send themselves messages:

```
#include <iostream>
#include <future>

#include <mqtt/client.h>

std::atomic<uint64_t> tick(0);

std::string wait_for_response(int thread_num) {
    mqtt::connect_options opts;
    opts.set_keep_alive_interval(20);
    opts.set_clean_session(true);

    tick++;

    std::string topic = "/test" + std::to_string(tick);

    mqtt::client cli("localhost", std::string("client"));
    cli.connect(opts);

    cli.subscribe(topic);

    std::string message = "this is a test message";
    cli.publish(topic, message.c_str(), message.length(), 2, false);

    cli.start_consuming();

    auto msg = cli.consume_message();
    std::string payload = msg.get()->get_payload_str();
    std::cout << "thread returned " << thread_num  << ", "<< payload << std::endl;

    cli.stop_consuming();
    cli.disconnect();

    return payload;
}

int main() {

    while (1) {
        std::future<std::string> f1 = std::async(std::launch::async, []{return wait_for_response(1);});
        std::future<std::string> f2 = std::async(std::launch::async, []{return wait_for_response(2);});
        std::future<std::string> f3 = std::async(std::launch::async, []{return wait_for_response(3);});

        f1.wait();
        f2.wait();
        f3.wait();

        std::cout << "async calls returned" << std::endl;
    }

    return 0;
}
```

If I am doing anything wrong, pointers would be appreciated. I guess this means that this library is not threadsafe?

**Comment by fpagliughi** [comment]
Both the C and C++ clients are thread safe. You should be able to use a single client from multiple threads, though that usually only makes sense for multiple threads publishing to a single client.

You should certainly be able to create multiple threads, each with its own client.

One problem I see is that you're using the same name for each client identifier, as "client". MQTT requires client ID's to be unique. You could use the `tick` value to create a unique name string for each client, or you can just use an empty client name, "", in which case the broker will assign a unique name to each client.

The broker might be rejecting the second and third connect attempts. Or it could possibly be dropping the first when the second connects, etc.

In a real system you should check that the connect succeeds, and handle a failure appropriately. You should normally also monitor for disconnects. But for your test, for now, just try using a blank client ID, "", and see if that fixes it.

If it does, it becomes a question of the proper behavior for `consume_message()`. For a synchronous client, I suppose it should return a failure when the client gets disconnected. But the application might want to be handling automatic reconnects in the background and let the main thread continue.

**Comment by jeremyherbert** [comment]
As I should have known, the bug was me :)

Yes, changing to "" fixes it. Perhaps on disconnect the `consume_message` function can throw an exception?

**Comment by fpagliughi** [comment]
Yes, this is confusing. `consume_message()` should definitely do _something_ if the client gets disconnected. I will keep this issue open, but perhaps rename it to make sure it stays on the TODO list.

**Comment by fpagliughi** [comment]
Oh, this was fixed long ago, (14dd81bff69b34bd5573231926602b4250e48ffe) and I forgot to close the issue. On the loss of connection an empty message is put in the queue.
